### DISCUSSION TITLE
can't access files from public folder

### DISCUSSION THREAD

**Discussion by hedihadi** [vercel/community#5094] [Help] [answered]
i have a public folder that contains a some js folders and some images, my structure looks like
```
root
--app.py
--vercel.json
--public
----result.png
----resposne.png
----src/jsFiles....
```

and this is my vercel.json file
```
{
 "builds": [{ "src": "app.py", "use": "@vercel/python" }],
 "routes": [
 { "src": "/(.*)", "dest": "app.py" },

           ]
}
```

but i can't access result.png or anything inside the public folder, why is that?

**Comment by SawkaDev** [comment, marked as answer]
@hedihadi have you had any luck fixing this?

**Reply by SawkaDev** [reply]
I got mine to work by doing the following, maybe try something similar

Answer for anyone who wants to figure out how to access a file stores in the /public folder in vercel deployment. Just do this

If your file is in `/public/assets` you just get it simply by the following:

`path.join(process.cwd(),"public","assets",FILENAME);`

**Reply by yousseframy24** [reply]
Where do I put this line can you add a full code snippet?

**Reply by alabbas-ali** [reply]
trying to read a file in the public folder, using :
```
const localFilePath = path.join(
            process.cwd(),
            process.env.NODE_ENV === 'production' ? 'static' : 'public',
            'GeoLite2-Country.mmdb'
        )
   ```     
        in the API function, I am still not able to access the file. there is always an error
        
         'no such file or directory, open '/var/task/static/GeoLite2-Country.mmdb'

**Reply by amyegan** [reply]
Like Sliffcak suggested, you can use `path.join()` to build the appropriate file path relative to the the root of your repo. There's an example in our guide about using files in serverless functions: https://vercel.com/guides/how-can-i-use-files-in-serverless-functions

**Reply by alabbas-ali** [reply]
@amyegan 

You can see already for the code I submitted in the comment that I am already using the `path.join()`. Also, your guide does not explain where the folder 'files' is located. 

Please look at the final path I am getting after using the `path.join()`. it is not a relative path at all.

**Reply by amyegan** [reply]
I believe the ternary operator in your example is part of the problem. Keep in mind that it's building a file path, not a URL. If the path to file from the root of the repo is `/static/GeoLite2-Country.mmdb`, then you can get the path using 

```javascript
const localFilePath = path.join(
    process.cwd(),
    'static',
    'GeoLite2-Country.mmdb'
);
```

A created a test app based on the guide to show a more complete example. I have `/files/test.json` and I can read it in my serverless function as shown in the guide.

<img width="732" alt="" src="https://github.com/vercel/community/assets/8247105/45b9f11d-69b8-45a8-9197-e58a5b5fad81">

**Comment by mpandzo** [comment]
@amyegan from what I was able to determine if you want to load a directory or file using fs.readdir or fs.readFile with Next.js 14 you must include it in the experimental outputFileTracingIncludes setting OR you must do the reading from within a Server Component. 

On the other hand I was not able to have the file or folder reading done in a library file which is then imported into a Server Action as it was not getting traced by Vercel Nft (https://github.com/vercel/nft).

Can you confirm this is the expected behaviour?

For anyone looking at this, an example of doing the reading from within a Server Component can be found here https://vercel.com/guides/loading-static-file-nextjs-api-route.

The config example of including a 'data' folder via outputFileTracingIncludes can be found below while the docs are here:
https://nextjs.org/docs/pages/api-reference/next-config-js/output

```
const nextConfig = {
  experimental: {
    outputFileTracingIncludes: {
      '/': ['./src/app/data/**/*'],
    },
  },
};
```
